src/main/java/org/opensearch/ad/feature/CompositeRetriever.java

```diff
@@ -0,0 +1,364 @@
+                            },
+                            "doc_count": 1,
+                            "the_max": {
+                                "value": -38.0
+                            },
+                            "the_min": {
+                                "value": -38.0
+                            }
+                        }
+                    ]
+               }
+             }
+             */
+            for (Bucket bucket : composite.getBuckets()) {
+                Optional<double[]> featureValues = parseBucket(bucket, anomalyDetector.getEnabledFeatureIds());
+                // bucket.getKey() returns a map of categorical field like "host" and its value like "server_1"
+                if (featureValues.isPresent() && bucket.getKey() != null) {
+                    results.put(Entity.createEntityByReordering(anomalyDetector.getDetectorId(), bucket.getKey()), featureValues.get());
+                }
+            }
+
+            totalResults += results.size();
+
+            afterKey = composite.afterKey();
+            return new Page(results);
+        }
+
+        private void updateCompositeAfterKey(SearchResponse r, SearchSourceBuilder search) {
+            CompositeAggregation composite = getComposite(r);
+
+            if (composite == null) {
+                throw new IllegalArgumentException(String.format(Locale.ROOT, "Empty resposne: %s", r));
+            }
+
+            updateSourceAfterKey(composite.afterKey(), search);
+        }
+
+        private void updateSourceAfterKey(Map<String, Object> afterKey, SearchSourceBuilder search) {
+            AggregationBuilder aggBuilder = search.aggregations().getAggregatorFactories().iterator().next();
+            // update after-key with the new value
+            if (aggBuilder instanceof CompositeAggregationBuilder) {
+                CompositeAggregationBuilder comp = (CompositeAggregationBuilder) aggBuilder;
+                comp.aggregateAfter(afterKey);
+            } else {
+                throw new IllegalArgumentException(
+                    String.format(Locale.ROOT, "Invalid client request; expected a composite builder but instead got {}", aggBuilder)
+                );
+            }
+        }
+
+        private boolean shouldRetryDueToEmptyPage(SearchResponse response) {
+            CompositeAggregation composite = getComposite(response);
+            // if there are no buckets but a next page, go fetch it instead of sending an empty response to the client
+            return composite != null && composite.getBuckets().isEmpty() && composite.afterKey() != null && !composite.afterKey().isEmpty();
+        }
+
+        CompositeAggregation getComposite(SearchResponse response) {
+            if (response == null || response.getAggregations() == null) {
+                return null;
+            }
+            Aggregation agg = response.getAggregations().get(AGG_NAME_COMP);
+            if (agg == null) {
```

Comment:
Consider using Optional to enforece the calller to handle the null case?

Reply:
yeah, changed to Optional.